rust/bridge/shared/src/node/convert.rs

```diff
@@ -632,47 +631,170 @@ pub(crate) unsafe fn extend_lifetime<'a, 'b: 'a, T: ?Sized>(some_ref: &'a T) ->
+    /// Factors out operations that differ between mutable and immutable bridge handles.
+    type Strategy: BridgeHandleStrategy<Self>;
+}
+
+/// Factors out operations that differ between mutable and immutable bridge handles.
+///
+/// See the [`BridgeHandle`] trait.
+pub trait BridgeHandleStrategy<T> {
+    /// If the Rust object needs any additional wrapping before being boxed by Neon,
+    /// this type can be used.
+    type JsBoxContents: From<T> + Send + 'static;
+}
+
+/// A phantom type for immutable bridge handles.
+///
+/// See [`BridgeHandleStrategy<T>`].
+pub struct Immutable<T>(std::marker::PhantomData<T>);
+
+/// A phantom type for mutable bridge handles.
+///
+/// See [`BridgeHandleStrategy<T>`].
+pub struct Mutable<T>(std::marker::PhantomData<T>);
+
+/// Immutable handles need no extra boxing.
+impl<T: BridgeHandle> BridgeHandleStrategy<T> for Immutable<T> {
+    type JsBoxContents = T;
+}
+
+/// Mutable handles use a [`RefCell`] for safety.
+impl<T: BridgeHandle> BridgeHandleStrategy<T> for Mutable<T> {
+    type JsBoxContents = RefCell<T>;
+}
+
+/// Shorthand for `T::Strategy::JsBoxContents` when `T: `[`BridgeHandle`], which Rust can't always
+/// resolve.
+type BoxContentsFor<T> = <<T as BridgeHandle>::Strategy as BridgeHandleStrategy<T>>::JsBoxContents;
+
+impl<'a, T: BridgeHandle> ResultTypeInfo<'a> for T {
+    type ResultType = JsValue;
+    fn convert_into(self, cx: &mut impl Context<'a>) -> NeonResult<Handle<'a, Self::ResultType>> {
+        Ok(cx
+            .boxed(DefaultFinalize(BoxContentsFor::<T>::from(self)))
+            .upcast())
+    }
+}
+/// Accesses a boxed Rust value using a custom reference (such as [`std::cell::RefMut`]).
+///
+/// Once created, the value can be accessed using the usual [`Deref`] and [`DerefMut`] operations.
+///
+/// Also stores the JavaScript handle that owns the JsBox to ensure that the Rust value is kept
+/// alive.
+pub struct BoxedBorrow<'a, Borrow: Deref + 'a>
+where
+    Borrow::Target: BridgeHandle,
+{
+    _owner: Handle<'a, DefaultJsBox<BoxContentsFor<Borrow::Target>>>,
+    borrow: Borrow,
+}
+
+impl<'a, Borrow: Deref + 'a> BoxedBorrow<'a, Borrow>
+where
+    Borrow::Target: BridgeHandle,
+{
+    /// Creates a BoxedBorrow by accessing `wrapper`, assuming it does in fact reference a boxed
+    /// Rust value under the `_nativeHandle` property.
+    ///
+    /// The `borrow` callback will be used to get an active reference for the Rust value.
+    /// For example, if the "box contents" for this handle use a [`RefCell`], `borrow` might
+    /// be [`RefCell::borrow_mut`] to get a mutable reference out. For a plain value, `borrow`
+    /// can be the identity function.
+    pub fn from_wrapper(
+        cx: &mut impl Context<'a>,
+        wrapper: Handle<'a, JsObject>,
+        borrow: fn(&'a BoxContentsFor<Borrow::Target>) -> Borrow,
+    ) -> NeonResult<Self> {
+        let boxed_value: Handle<'a, DefaultJsBox<BoxContentsFor<Borrow::Target>>> =
```

Comment:
This would still be more clear to me if it were "js_boxed_bridge_handle" to make it clear it's a JsBox and not a normal Box, and we know the contents to be a BridgeHandle.  I know that's in the type, but the type is pretty complex.